test/distributed/fsdp/test_fsdp_use_orig_params.py

```diff
@@ -0,0 +1,875 @@
+            optim_orig_params.step()
+        self._check_fsdp_parameter_parity(fsdp_model, fsdp_model_orig_params)
+
+
+class TestFSDPUseOrigParamsParamAccess(FSDPTest):
+    """Tests original parameter access."""
+
+    @property
+    def world_size(self):
+        # Force a world size of 2 since the tests hard code to the FSDP
+        # sharding strategy to check sharded parameter parity
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_access_params_after_forward(self):
+        """
+        Tests that accessing the original parameters after the forward but
+        before the backward. Notably, this is not supported when
+        ``use_orig_params=False``. However, for ``True``, FSDP exposes the
+        (flattened) sharded original parameters, making it possible.
+        """
+        self.run_subtests(
+            {
+                "sharding_strategy": [
+                    ShardingStrategy.NO_SHARD,
+                    ShardingStrategy.FULL_SHARD,
+                    ShardingStrategy.SHARD_GRAD_OP,
+                ],
+            },
+            self._test_access_params_after_forward,
+        )
+
+    def _test_access_params_after_forward(
+        self,
+        sharding_strategy: ShardingStrategy,
+    ):
+        class Model(nn.Module):
+            def __init__(self):
+                super().__init__()
+                torch.manual_seed(42)
+                # 5 * 5 = 25 numel -> pad to 26 -> 13 on each rank
+                self.lin1 = nn.Linear(5, 5, bias=False)
+                # 5 * 7 + 7 = 42 numel -> no pad -> 21 on each rank
+                # 21 of weight on rank 0; 14 of weight and 7 of bias on rank 1
+                self.lin2 = nn.Linear(5, 7)
+
+            def forward(self, x: torch.Tensor) -> torch.Tensor:
+                z = self.lin1(x)
+                z = nn.functional.relu(z)
+                z = self.lin2(z)
+                return z
+
+            def get_input(self, device: torch.device) -> Tuple[torch.Tensor, ...]:
+                return (torch.randn((2, 5)).to(device),)
+
+            def get_loss(self, inp, out):
+                return out.sum()
+
+        def check_parameter_parity(ddp_model, fsdp_model):
+            assert self.rank in (
+                0,
+                1,
+            ), f"Expects world size of 2 but got {self.world_size}"
+            for (n1, p1), (n2, p2) in zip(
+                ddp_model.module.named_parameters(),
+                fsdp_model.named_parameters(),
+            ):
+                self.assertEqual(n1, clean_tensor_name(n2))
+                if sharding_strategy == ShardingStrategy.NO_SHARD:
+                    # For `NO_SHARD`, do nothing since the original parameters
+                    # are unflattened
+                    pass
+                # Otherwise, case on the parameter (see the model definition)
+                elif n1 == "lin1.weight":
+                    if self.rank == 0:
+                        p1 = p1.flatten()[:13]
```

Comment:
this test seems a little bit brittle and seems to make assumptions about the FSDP sharding details? But I guess there's no better way to test original parameters to be the same as the parameters that use_orig_params=False returns. 

Is there no FSDP API (even private) to get the sharded portion of a parameter with use_orig_params=False?